### PR TITLE
Belos & Amesos2: changes following Tpetra_UVM_Removal branch

### PR DESCRIPTION
**Pull request by iyamazaki**
@trilinos/tpetra, @trilinos/belos, and @trilinos/amesos2 

## Motivation
This PR contains changes needed for Belos & Amesos2, following up the recent changes in Tpetra to remove UVM dependencies [(PR#9038)](https://github.com/trilinos/Trilinos/pull/9038)

## Related Issues

Closes [Issue#8761](https://github.com/trilinos/Trilinos/issues/8761)

## Testing
Tested on Vortex with/without UVM. For Amesos2, we tested with KLU2 and SuperLU, using MPI enabled.

### CODE REVIEW

**Comment by jennloe** [comment]
@hkthorn ??

**Comment by csiefer2** [comment]
As a Tpetroid, the Belos changes look fine to me!

**Review by lucbv — APPROVED** [review]
For some reason the auto-tester is not testing this so I am approving hoping this will make the auto-tester run. The AT:AUTO-MERGE is not used so this should not cause a problem

**Comment by hkthorn** [comment]
@csiefer2 I added you so you can approve this merge request.

**Comment by jhux2** [comment]
@lucbv Ross has already reported more AT woes on the usual issue.

**Comment by csiefer2** [comment]
@jhux2 I really need to write that autoretester...

**Review by csiefer2 — APPROVED** [review]

File: packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp:139]: @iyamazaki thanks for the work on this PR, I have a couple questions on some Amesos2 changes:

1. There are several locations where routines like this `getCrs` are guarded based on enabled deprecated code but there is not an else branch, why does these need to be removed? I'm not involved with the Tpetra updates, it's not clear to me why/if this is necessary to coincide with those changes or just considered clean up
2. Can you confirm which solvers were enabled in testing, and was this PR tested with/without MPI? Some of the remnant non-Kokkos view routines may be necessary to support matrix inputs with non-contiguous GIDs, which are only tested with MPI enabled. If you could confirm the following KLU2, Basker and SuperLU tests ran and passed with these changes, that would help me feel comfortable with the removals (I think only KLU2 is enabled in PR testing):
https://github.com/trilinos/Trilinos/blob/a6d3274e2f75876e7d682e34f92b12dded96f014/packages/amesos2/test/solvers/Basker_UnitTests.cpp#L389
https://github.com/trilinos/Trilinos/blob/a6d3274e2f75876e7d682e34f92b12dded96f014/packages/amesos2/test/solvers/KLU2_UnitTests.cpp#L419
https://github.com/trilinos/Trilinos/blob/a6d3274e2f75876e7d682e34f92b12dded96f014/packages/amesos2/test/solvers/Superlu_UnitTests.cpp#L374
https://github.com/trilinos/Trilinos/blob/a6d3274e2f75876e7d682e34f92b12dded96f014/packages/amesos2/test/solvers/ShyLUBasker_UnitTests.cpp#L425

File: packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp:164]: Ditto

File: packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp:209]: Ditto

File: packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp:233]: Ditto

File: packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp:331]: Ditto

File: packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp:361]: Ditto

File: packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp:391]: Ditto

File: packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp
Comment (ndellingwood) [inline packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp:421]: Ditto

File: packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp
Comment (csiefer2) [reply packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp:139]: This is an *excellent* point.  The path I took in Ifpack2 is to use the deprecated Tpetra interfaces *only* when I was inheriting from a Tpetra object which had deprecated interfaces.  Since Amesos2 is not doing this, those interfaces can probably go.

Comment (iyamazaki) [reply packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp:139]: Thank you for the review, @ndellingwood!! The most of the changes for Amesos2 are because Tpetra is switching to return the local Crs as Kokkos::views instead of Teuchos::ArrayViews. There are public slides that summarize the MV changes, but I am not sure if there are slides about Crs changes. I can ask @kddevin, if they would help.

I have been testing with KLU2 and SuperLU, but not with Baskers, using MPI. I updated the motivation above, and the [Issue#8761](https://github.com/trilinos/Trilinos/issues/8761) with my Amesos2 configure.

Comment (iyamazaki) [reply packages/amesos2/src/Amesos2_MatrixAdapter_decl.hpp:139]: Thank you for the inputs, @csiefer2. For now, we kept the old interfaces, which use the deprecated Tpetra interfaces, as deprecated codes because I was not sure if any solvers or other codes use these old adapter interfaces. So, if they do, these solvers need to enable the Tpetra's deprecated code. But, I can just remove them, if that is better!

**Comment by brian-kelley** [comment]
@iyamazaki It looks like Amesos2_KLU2_def.hpp does this around line 520:
```
Util::get_ccs_helper_kokkos_view<
        MatrixAdapter<Matrix>,host_value_type_array,host_ordinal_type_array,host_ordinal_type_array>
        ::do_get(this->matrixA_.ptr(), host_nzvals_view_, host_rows_view_, host_col_ptr_view_,
            nnz_ret, ROOTED, ARBITRARY, this->rowIndexBase_);
```
but Matrix is an Epetra_CrsMatrix when the error happens, which explains why GO was int. I think ``get_cxs_helper_kokkos_view`` needs a separate codepath for Epetra objects.

**Comment by iyamazaki** [comment]
Thank you, @brian-kelley. That was very helpful. I think I forgot fixing the changes I made to the MatrixTraits for Epetra. I am still building to reproduce the errors, but will try fixing it.

**Comment by jennloe** [comment]
@trilinos/framework Could we please give @iyamazaki write access?  Looks like he is a member of the Belos team, but not Trilinos / Developers.

**Review by csiefer2 — APPROVED** [review]

**Comment by csiefer2** [comment]
> 
> 
> @trilinos/framework Could we please give @iyamazaki write access? Looks like he is a member of the Belos team, but not Trilinos / Developers.

Taken care of.

**Review by csiefer2 — DISMISSED** [review]

**Review by csiefer2 — APPROVED** [review]

File: packages/amesos2/src/Amesos2_TpetraMultiVecAdapter_def.hpp
Comment (brian-kelley) [inline packages/amesos2/src/Amesos2_TpetraMultiVecAdapter_def.hpp:205]: I don't think ``X_lcl_j_2d`` is necessary here since you already have ``contig_local_view_2d`` above.

File: packages/amesos2/src/Amesos2_TpetraMultiVecAdapter_def.hpp
Comment (brian-kelley) [inline packages/amesos2/src/Amesos2_TpetraMultiVecAdapter_def.hpp:208]: Could you use ``typename multivec_t::impl_scalar_type`` instead for readability?

File: packages/amesos2/src/Amesos2_TpetraMultiVecAdapter_def.hpp
Comment (brian-kelley) [inline packages/amesos2/src/Amesos2_TpetraMultiVecAdapter_def.hpp:476]: Same as above - could use multivec_t::impl_scalar_type

**Review by brian-kelley — CHANGES_REQUESTED** [review]

**Review by ndellingwood — APPROVED** [review]
Amesos2 changes look fine (a few remnant commented lines can be cleaned up later) and testing passes, thanks for the hard work @iyamazaki !

**Comment by iyamazaki** [comment]
Thank you for the comments, @brian-kelley. I have the changes, but I am thinking to push them as a separate PR (since auto-tester seems to be struggling right now). Or, I can create an issue with a list of changes if @ndellingwood also has some inputs. What do you think?

**Review by brian-kelley — APPROVED** [review]

**Comment by brian-kelley** [comment]
@iyamazaki That sounds fine. I just approved this, but my comments from the first review are still up there.

**Comment by william76** [comment]
@prwolfe @allevin 
I got asked about this PR... the approval boxes got checked yesterday around 8am after the testing passed but the pre-merge inspection flag seems to still be unset.

Is it possible that there are just a bunch of jobs in this tester's swim lane and we need to wait a bit more for it to cycle around or could something else be blocking it?

@csiefer2 Paul or Aaron may have more insight into the delay than I do... but in the past my experience has generally been that the tester needs to cycle around and do its check of the PR to set the final flags.  Still, you approved this about 26 hours ago which seems long... but builds also take a long time too so if there's other stuff pending.... 🤷🏻‍♂️

**Comment by prwolfe** [comment]
@william76 - we are currently struggling with the change in the proxy system and it's implications when run through python. Basically I can get the autotester to talk to GitHub or to talk with Jenkins, but not both as we need.

**Review by srajama1 — COMMENTED** [review]
Not sure why Amesos2 is still showing as locked, approving to see if that fixes it.

**Pull request merged**
